airflow/models/dag.py

```diff
@@ -1189,11 +1195,25 @@ def relative_fileloc(self) -> pathlib.Path:
         """File location of the importable dag 'file' relative to the configured DAGs folder."""
         path = pathlib.Path(self.fileloc)
         try:
-            return path.relative_to(settings.DAGS_FOLDER)
+            rel_path = path.relative_to(self.dag_processor_dags_folder or settings.DAGS_FOLDER)
+            if rel_path == pathlib.Path('.'):
+                return path
+            else:
+                return rel_path
         except ValueError:
             # Not relative to DAGS_FOLDER.
             return path
 
+    @property
+    def dag_processor_dags_folder(self):
```

Comment:
nit: How does it apply to `processor_subdir` field from e.g.DagModel? 
Doesn't it have the same meaning?
Maybe we should keep the name consistent?

Reply:
Hmm... that's a good question....  I don't know if it has the same meaning.  Is it correct to assume that processor_subdir will always be the dags folder of the processor?  If so, why didn't they call it dags folder in the first place?  What if there is sometimes an optimization to pass the actual dag file path as subdir to the dag processor so that it will just load that specific file instead of all dags in the dags folder?

Reply:
Actually... i think processor_subdir is something different.  I think that the idea is you can have multiple dag processors running, each looking at a different subdir, which is a subdir of dags_folder.  that sound right to you?

Reply:
from the looks of it, that's why e.g. you have a `--subdir` option in airflow dags-processor command (which can be different from DAGS_FOLDER) 
and then you have this logic
https://github.com/apache/airflow/blob/main/airflow/dag_processing/manager.py#L508-L509
this logic appears to say, "if i am a standalone dags processor, and i didn't find the dags in my path, then mark them as stale"

it does seem a bit weird though because then maybe if you change the dags processor subdirs later, perhaps some of those dags would never be returned by this query and therefore would never be deactivated?

Reply:
ah i guess you would know better than most ;) 
<img width="679" alt="image" src="https://user-images.githubusercontent.com/15932138/191809049-5b553a1a-d0b5-41e4-9dcc-7182ee21a127.png">
let me know your thoughts

Reply:
There's nothing that force the processors to look at subfolders of the dags folder -- you could have one looking at `/opt/dags` and another at `/home/airflow/airflow/dags` and it would "work"

Reply:
> There's nothing that force the processors to look at subfolders of the dags folder

Isn't that what the `--subdir` option does?

> you could have one looking at `/opt/dags` and another at `/home/airflow/airflow/dags` and it would "work"

not seeing the connection?

Reply:
I see that name may be a little confusing, but `--subdir` just points to the dag folder - the value can be absolute path and two different processors may point ot complete different locations.

So what you wrote above:

> Actually... i think processor_subdir is something different. I think that the idea is you can have multiple dag processors running, each looking at a different subdir, which is a subdir of dags_folder. that sound right to you?

is actually not true

Also

> it does seem a bit weird though because then maybe if you change the dags processor subdirs later, perhaps some of those dags would never be returned by this query and therefore would never be deactivated?

For that we have a new check in the scheduler:
https://github.com/apache/airflow/blob/main/airflow/jobs/scheduler_job.py#L1541

Reply:
> I see that name may be a little confusing, but --subdir just points to the dag folder - the value can be absolute path and two different processors may point ot complete different locations.

what do you mean "just points to the dag folder"? i have looked at the code and it does not seem that there's any logic e.g. to make the subdir agree with configured dags folder.  e.g. setting subdir on dag processor command does not force dags_folder to agree.  so fundamentally, they are two different things right?  the dag processor can run without them agreeing, no?  

if they are in completely different locations, i.e. not a "subdir" of the dags folder, then the concept of "relative location" of a dag file becomes kind of meaningless, and if you can't use relative loc, then the worker can't really rely on ser dag to know where a dag is located.

Reply:
> to make the subdir agree with configured dags folder. e.g. setting subdir on dag processor command does not force dags_folder to agree. so fundamentally, they are two different things right? the dag processor can run without them agreeing, no?

Right, there is no logic/check, we just pass --subdir to the Dag Processor and it parses the dags from this directory. 

>  if you can't use relative loc, then the worker can't really rely on ser dag to know where a dag is located.

Well, if you define different set of workers on different task queues then it may work, but I agree there is no support from Airflow perspective for such setup.

But note that dag processor seperation (and multiple dag processors) is just the first step towards Airflow multit-tenancy. At some point workers isolation will also be supported.

cc: @potiuk

Reply:
> Are we on the same page here :) ?

maybe sort of? :) i'm not sure what you're proposing though...

taking it back down to where the rubber meets the road here...  what i'm trying to do here is fix a bug.  currently if the DAGS_FOLDER of scheduler/processor differs from that of k8s executor then all tasks will fail to run.  i think this issue is the same one you say worked in 1.10 but was broken in 2.0  and 2.1 -- it's broken again in 2.4 and that's what i'm trying to fix.

this is easy to repro if you run scheduler in local virtualenv and run your k8s pods in docker desktop

the cause is the recent optimization to not "process" dags in `airflow tasks run --local` but to read from ser dag.  the pod definition remains like this
```
  - args:
    - airflow
    - tasks
    - run
    - example_bash_operator
    - runme_0
    - manual__2022-09-23T20:04:52.306715+00:00
    - --local
    - --subdir
    - DAGS_FOLDER
```
But when that gets converted to `tasks run --raw`  then the subdir becomes the full path (`fileloc`) of the dag _as it is located on the scheduler/processor`.  I think it breaks down because, our mechanism for handling different dags folders is `relative_fileloc`.  But currently the logic to calculate relative fileloc is based on the _current_ dags folder, which means, if the _current_ process has a different dags folder, it cannot be calculated!  So then it defaults to the full filepath as stored in serdag and therefore it fails to find the dag.

So what do we do about it?  The relatively straightforward solution is to use the dags_folder of serializer instead of dags_folder of current process when calculating relative fileloc (or equivalently, you could actually store the relative fileloc when serializing).  Then the actual path can always be reassembled, provided that the dag _actually is_ relative to the dags_folder in all envs (and has the same relative path).

So that's why I added dag_processor_dags_folder.

But @mhenc is suggesting, I think, to just use `processor_subdir` instead.  But the problem I see with that is, processor_subdir may not always equal dags_folder.  And in the case of multiple processors in subdirs of dags folder, then this would not work I think.  The only way it would work is if you had a 1-1 relationship between dags_folder of worker and processor_subdir of processor.  But in that scenario, why not just set the DAGS_FOLDER of your processor to be the subdir?

Fundamentally, I think that currently airflow assumes everything must be in the dags folder, and that dag processors, to the extent they have different paths (subdir) they are _subdirs_ of the dags folder.  This alone seems QED for the notion that processor_subdir !== dags_folder.  I think as we get closer to multitenancy, it seems likely we'll need to blow up the dags_folder concept; if it's no longer true that there is _one_ dags folder then the code and language need to reflect that.  There may be many dags folders.  "relative_fileloc" won't necessarily make sense as a concept (unless it's redefined to mean relative to the dags folder of this "sub-instance").  But for now, the two (processor_subdir) and dags folder seem to be two different things, and DAGS_FOLDER seems to have the best likelihood of giving a reliable answer for `relative_fileloc` so it seems worth storing separately.

THAT SAID, perhaps we should make it a "private" attribute, so as not to increase the backcompat surface unnecessarily?

WDYT?

Reply:
@potiuk 

> I believe (please correct me if I misundderstood it @dstandish) that this PR here was very similar to what the https://github.com/apache/airflow/pull/16860 change for @ashb was. But while Ashb wanted to resurrect the relative approach working for the main DAGS folder, but this one was to bring it to the standalone DAG processor.

Not quite.... the "change" here is not really a change, it's fixing the regression in 2.4 of that same exact feature.  It's not specific to dag processor.  This issue appears even just using scheduler with virtualenv and k8s exec --- no dags processor process required.